apiserver/facades/controller/caasapplicationprovisioner/provisioner.go

```diff
@@ -185,16 +187,60 @@ func NewStateCAASApplicationProvisionerAPI(ctx facade.ModelContext) (*APIGroup,
 
 	apiGroup := &APIGroup{
 		PasswordChanger:    common.NewPasswordChanger(st, common.AuthFuncForTagKind(names.ApplicationTagKind)),
-		LifeGetter:         common.NewLifeGetter(st, lifeCanRead),
 		AgentEntityWatcher: common.NewAgentEntityWatcher(st, ctx.Resources(), common.AuthFuncForTagKind(names.ApplicationTagKind)),
 		charmInfoAPI:       commonCharmsAPI,
 		appCharmInfoAPI:    appCharmInfoAPI,
+		lifeCanRead:        lifeCanRead,
 		API:                api,
 	}
 
 	return apiGroup, nil
 }
 
+// Life returns the life status of every supplied app or unit, where available.
+func (a *APIGroup) Life(ctx context.Context, args params.Entities) (params.LifeResults, error) {
```

Comment:
How annoying this is almost the same as the uniter agent facade. Oh well, we can deal with this in an API redesign.

Reply:
So yeah, the common API plugins like LifeGetter, Remove, EnsureDeader etc  were meant to stop the code duplication on the facade structs. However, we ran have run into issues around versioning etc. One option we discussed at the time was to have an instance of the common struct instead of embedding it, and then thunk to it via  a top level method on the facade.
However, I wanted just to do it this way to start with - yes a bit of duplication etc. But it's minimal and is essentially just iterating over some api args, parsing a tag, doing an api call, returning result. Let's keep everything simple and flat to start with, and once all the facades have the common plugins refactored out and sittling on top of the new services, then we can look at optimisation, rather than doing it now (early) and maybe missing something.